posthog/async_migrations/migrations/0005_person_collapsed_by_version.py

```diff
@@ -0,0 +1,252 @@
+                database=AnalyticsDBMS.CLICKHOUSE,
+                sql=f"""
+                    INSERT INTO {TEMPORARY_TABLE_NAME}
+                    SELECT *
+                    FROM {PERSON_TABLE}
+                """,
+                rollback=f"TRUNCATE TABLE IF EXISTS {TEMPORARY_TABLE_NAME} ON CLUSTER '{settings.CLICKHOUSE_CLUSTER}'",
+                timeout_seconds=2 * 24 * 60 * 60,  # two days
+            ),
+            AsyncMigrationOperation(
+                fn=self.copy_persons_from_postgres, rollback_fn=lambda _: self.unset_highwatermark()
+            ),
+            AsyncMigrationOperationSQL(
+                database=AnalyticsDBMS.CLICKHOUSE,
+                sql=f"DROP TABLE IF EXISTS {TEMPORARY_PERSON_MV} ON CLUSTER '{settings.CLICKHOUSE_CLUSTER}'",
+                rollback=None,
+            ),
+            AsyncMigrationOperationSQL(
+                database=AnalyticsDBMS.CLICKHOUSE,
+                sql=f"DROP TABLE IF EXISTS person_mv ON CLUSTER '{settings.CLICKHOUSE_CLUSTER}'",
+                rollback=None,
+            ),
+            AsyncMigrationOperationSQL(
+                database=AnalyticsDBMS.CLICKHOUSE,
+                sql=f"""
+                    RENAME TABLE
+                        {PERSON_TABLE_NAME} to {BACKUP_TABLE_NAME},
+                        {TEMPORARY_TABLE_NAME} to {PERSON_TABLE_NAME}
+                    ON CLUSTER '{settings.CLICKHOUSE_CLUSTER}'
+                """,
+                rollback=f"""
+                    RENAME TABLE
+                        {PERSON_TABLE_NAME} to {FAILED_PERSON_TABLE_NAME},
+                        {BACKUP_TABLE_NAME} to {PERSON_TABLE_NAME}
+                    ON CLUSTER '{settings.CLICKHOUSE_CLUSTER}'
+                """,
+            ),
+            AsyncMigrationOperationSQL(database=AnalyticsDBMS.CLICKHOUSE, sql=PERSONS_TABLE_MV_SQL, rollback=None),
+            AsyncMigrationOperation(fn=optimize_table_fn),
+        ]
+
+    def new_table_engine(self):
+        engine = ReplacingMergeTree("person", ver="version")
+        # :TRICKY: Zookeeper paths need to be unique each time we run this migration, so generate a unique prefix.
+        engine.set_zookeeper_path_key(now().strftime("am0005_%Y%m%d%H%M%S"))
+        return engine
+
+    @cached_property
+    def pg_copy_target_person_id(self) -> int:
+        # :TRICKY: We calculate the last ID to copy at the start of migration once and cache it.
+        #    If the migration gets e.g. restarted, this is recalculated.
+        try:
+            return Person.objects.latest("id").id
+        except Person.DoesNotExist:
+            return -1
+
+    def get_pg_copy_highwatermark(self) -> int:
+        highwatermark = get_client().get(REDIS_HIGHWATERMARK_KEY)
+        return int(highwatermark) if highwatermark is not None else 0
+
+    def unset_highwatermark(self) -> None:
+        get_client().delete(REDIS_HIGHWATERMARK_KEY)
+
+    def copy_persons_from_postgres(self, query_id: str):
```

Comment:
@yakkomajuri I need your help around planning this step (at least on cloud).

Problem: take a darn long time - at minimum 5 days on cloud and that is likely optimistic. I designed it in a way that's resumable but not sure how this will behave operationally if celery task gets killed. Note that the new cloud infra @hazzadous / @guidoiaquinti are pushing live might affect operations here as well.

Reply:
We discussed this live - basically if we're not forcing this into the release that's fine.

I will also change the order of operations so postgres copying is done after everything else is live \o/